examples/cfd/swe_distributed_gnn/pde_dataset.py

```diff
@@ -0,0 +1,147 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+import torch
+
+from math import ceil
+
+from shallow_water_equations import ShallowWaterSolver
+
+
+class PdeDataset(torch.utils.data.Dataset):
+    """Custom Dataset class for PDE training data"""
+
+    def __init__(
+        self,
+        dt,
+        nsteps,
+        dims=(384, 768),
+        pde="shallow water equations",
+        initial_condition="random",
+        num_examples=32,
+        device=torch.device("cpu"),
+        normalize=True,
+        rank=0,
+        stream=None,
+        dtype=torch.float32,
+    ):
+        self.dtype = dtype
+
+        self.num_examples = num_examples
+        self.device = device
+        self.stream = stream
+        self.rank = rank
+
+        self.nlat = dims[0]
+        self.nlon = dims[1]
+
+        # number of solver steps used to compute the target
+        self.nsteps = nsteps
+        self.normalize = normalize
+
+        if pde == "shallow water equations":
+            lmax = ceil(self.nlat / 3)
+            mmax = lmax
+            dt_solver = dt / float(self.nsteps)
+            self.solver = (
+                ShallowWaterSolver(
+                    self.nlat,
+                    self.nlon,
+                    dt_solver,
+                    lmax=lmax,
+                    mmax=mmax,
+                    grid="equiangular",
+                )
+                .to(self.device)
+                .float()
+            )
+        else:
+            raise NotImplementedError
+
+        self.set_initial_condition(ictype=initial_condition)
+
+        inp0, tar0 = self._get_sample()
+        self.inp_shape = inp0.shape
+        self.tar_shape = tar0.shape
+
+        if self.normalize:
+            self.inp_mean = torch.mean(inp0, dim=(-1, -2)).reshape(-1, 1, 1)
+            self.inp_var = torch.var(inp0, dim=(-1, -2)).reshape(-1, 1, 1)
+
+    def __len__(self):
+        length = self.num_examples if self.ictype == "random" else 1
+        return length
+
+    def set_initial_condition(self, ictype="random"):
+        self.ictype = ictype
+
+    def set_num_examples(self, num_examples=32):
+        self.num_examples = num_examples
+
+    def _get_sample(self):
+        if self.ictype == "random":
+            inp = self.solver.random_initial_condition(mach=0.2)
+        elif self.ictype == "galewsky":
+            inp = self.solver.galewsky_initial_condition()
```

Comment:
I would assume so. Note that most parts of this example have been provided by Mohammad and are more or less taken from the FNO example in torch-harmonics. (this includes e.g. also the naming of files etc.)
Should we modify the example then to make it more reasonable? We can always remove things like e.g. the galewsky initial condition and streamline other things, too.
I mainly focused on "distributing" the given example.

Reply:
I think it is okay to leave it as is. Maybe worth adding a small comment here just to clarify this is only an example and not a fully fleshed out recipe?

Reply:
Looked a bit into this, this "static" initial condition is one that is often used as a reference of some kind for testing things. I added a comment that this can be used for valuation of trained models for instance.